### PR TITLE
[ODS-4125] Support Download of Xsd Metadata From Deployed ODS/API

### PR DESCRIPTION
**Pull request by jamessmckay**

### CODE REVIEW

**Review by blmeyers — CHANGES_REQUESTED** [review]

File: Application/EdFi.Ods.Api/Controllers/VersionController.cs [hunk 1]
```diff
@@ -44,85 +47,96 @@ public class VersionController : ControllerBase
...
                 var instance = "{instance}";
 
-                bool isYearSpecific = _apiSettings.GetApiMode().Equals(ApiMode.YearSpecific);
-                bool isInstanceYearSpecific = _apiSettings.GetApiMode().Equals(ApiMode.InstanceYearSpecific);
+                bool isYearSpecific = _apiSettings.GetApiMode().Equals(ApiMode.YearSpecific)
+                                      || _apiSettings.GetApiMode().Equals(ApiMode.InstanceYearSpecific);
 
-                //This also involves year itself, thus year needs to be true
-                if (isInstanceYearSpecific) isYearSpecific = true;
+                bool isInstanceYearSpecific = _apiSettings.GetApiMode().Equals(ApiMode.InstanceYearSpecific);
```
[the remark this answers is not in the document]
Comment (jamessmckay) [reply Application/EdFi.Ods.Api/Controllers/VersionController.cs:82]: done

File: Application/EdFi.Ods.Api/Middleware/XsdMetadataFileMiddleware.cs [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+        public async Task InvokeAsync(HttpContext context, RequestDelegate next)
+        {
+            if (context.Request.Method != HttpMethods.Get)
+            {
+                await next(context);
+                return;
+            }
+
+            var routeMatcher = new RouteMatcher();
```
[the remark this answers is not in the document]
Comment (jamessmckay) [reply Application/EdFi.Ods.Api/Middleware/XsdMetadataFileMiddleware.cs:40]: I think that maybe a viable path worth exploring, I created a ticket ODS-4787

File: Application/EdFi.Ods.Api/Middleware/XsdMetadataFileMiddleware.cs [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+            var xsdFileInformationByUriSegment = _xsdFileInformationProvider.XsdFileInformationByUriSegment();
+
+            if (!xsdFileInformationByUriSegment.ContainsKey(schema))
+            {
+                context.Response.StatusCode = StatusCodes.Status404NotFound;
+                return;
+            }
+
+            string assemblyName = xsdFileInformationByUriSegment[schema].AssemblyName;
+            string fullQualifiedFileName = $"{assemblyName}.Artifacts.Schemas.{routeValues["file"]}.xsd";
```
[the remark this answers is not in the document]
Comment (jamessmckay) [reply Application/EdFi.Ods.Api/Middleware/XsdMetadataFileMiddleware.cs:59]: ODS-4788

File: Application/EdFi.Ods.Api/Models/XsdFileInformation.cs [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+// The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
+// See the LICENSE and NOTICES files in the project root for more information.
+
+using System.Text;
+using EdFi.Common.Extensions;
+using EdFi.Ods.Common;
+
+namespace EdFi.Ods.Api.Models
+{
+    public class XsdFileInformation
```
[the remark this answers is not in the document]
Comment (jamessmckay) [reply Application/EdFi.Ods.Api/Models/XsdFileInformation.cs:12]: done

File: Application/EdFi.Ods.Api/Providers/EmbeddedFileProvider.cs [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+    {
+        private readonly IAssembliesProvider _assembliesProvider;
+
+        public EmbeddedFileProvider(IAssembliesProvider assembliesProvider)
+        {
+            _assembliesProvider = assembliesProvider;
+        }
+
+        public string File(string assemblyName, string fullyQualifiedName)
+            => _assembliesProvider.Get(assemblyName).ReadResource(fullyQualifiedName);
```
[the remark this answers is not in the document]
Comment (jamessmckay) [reply Application/EdFi.Ods.Api/Providers/EmbeddedFileProvider.cs:21]: Converted to streams, if the file is not found it returns a null stream. if the assembly is not found it returns a null stream

File: Application/EdFi.Ods.Api/Providers/IXsdFileInformationProvider.cs [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using EdFi.Ods.Api.Models;
+
+namespace EdFi.Ods.Api.Providers
+{
+    public interface IXsdFileInformationProvider
+    {
+        IDictionary<string, XsdFileInformation> XsdFileInformationByUriSegment();
```
[the remark this answers is not in the document]
Comment (jamessmckay) [reply Application/EdFi.Ods.Api/Providers/IXsdFileInformationProvider.cs:15]: done

File: Application/EdFi.Ods.Api/Providers/XsdFileInformationProvider.cs [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+        {
+            _assembliesProvider = assembliesProvider;
+            _schemaNameMapProvider = schemaNameMapProvider;
+
+            _xsdFileInformationByUriSegment = new Lazy<IDictionary<string, XsdFileInformation>>(ParseAssemblies);
+        }
+
+        public IDictionary<string, XsdFileInformation> XsdFileInformationByUriSegment() => _xsdFileInformationByUriSegment.Value;
+
+        public Task<XsdFileInformation> XsdFileInformationAsync(string schema, CancellationToken cancellationToken = default)
```
[the remark this answers is not in the document]
Comment (jamessmckay) [reply Application/EdFi.Ods.Api/Providers/XsdFileInformationProvider.cs:39]: done

File: Application/EdFi.Ods.Api/Providers/XsdFileInformationProvider.cs [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+
+                _logger.Debug(assemblySchemaInformation);
+
+                assemblySchemaInformationByUriSegment.Add(
+                    assemblySchemaInformation.SchemaNameMap.UriSegment, assemblySchemaInformation);
+            }
+
+            return assemblySchemaInformationByUriSegment;
+
+            static bool IsIncluded(string name)
```
[the remark this answers is not in the document]
Comment (jamessmckay) [reply Application/EdFi.Ods.Api/Providers/XsdFileInformationProvider.cs:107]: fixed the where clause and removed the list

File: Application/EdFi.Ods.Api/Startup/OdsStartupBase.cs [hunk 9]
```diff
@@ -90,6 +92,16 @@ public void ConfigureServices(IServiceCollection services)
...
+            services.AddSingleton<IActionContextAccessor, ActionContextAccessor>();
+
+            services.AddScoped(
+                serviceProvider =>
+                {
+                    var actionContext = serviceProvider.GetRequiredService<IActionContextAccessor>().ActionContext;
+                    var factory = serviceProvider.GetRequiredService<IUrlHelperFactory>();
+                    return factory.GetUrlHelper(actionContext);
+                });
+
```
[the remark this answers is not in the document]
Comment (jamessmckay) [reply Application/EdFi.Ods.Api/Startup/OdsStartupBase.cs:104]: this seems to be the preferred way of generating links, with the usage of reverse proxies. this tool applies the correct resolution for endpoints.

File: Application/EdFi.Ods.Features/XsdMetadata/XsdMetadataController.cs [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+            var xsdFileInformationByUriSegment = _xsdFileInformationProvider.XsdFileInformationByUriSegment();
+
+            if (!xsdFileInformationByUriSegment.Keys.Any())
+            {
+                return Ok(new Dictionary<string, object>[0]);
+            }
+
+            return Ok(
+                xsdFileInformationByUriSegment.Keys
+                    .Select(uriSegment => xsdFileInformationByUriSegment[uriSegment])
```
[the remark this answers is not in the document]
Comment (jamessmckay) [reply Application/EdFi.Ods.Features/XsdMetadata/XsdMetadataController.cs:55]: this has been corrected via the xsdfileprovider refactor above

File: Application/EdFi.Ods.Features/XsdMetadata/XsdMetadataController.cs [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+                        xsdFileInformation =>
+                            new Dictionary<string, object>(StringComparer.InvariantCultureIgnoreCase)
+                            {
+                                ["description"] = xsdFileInformation.IsCore()
+                                    ? $"Core schema ({xsdFileInformation.SchemaNameMap.LogicalName}) files for the data model"
+                                    : $"Extension {xsdFileInformation.SchemaNameMap.LogicalName} blended with Core schema files for the data model",
+                                ["name"] = xsdFileInformation.SchemaNameMap.UriSegment,
+                                ["version"] = xsdFileInformation.Version,
+                                ["files"] = new Uri(
+                                    $"{_urlHelper.ActionLink("Get", ControllerContext.ActionDescriptor.ControllerName)}/{xsdFileInformation.SchemaNameMap.UriSegment}/files/")
```
[the remark this answers is not in the document]
Comment (jamessmckay) [reply Application/EdFi.Ods.Features/XsdMetadata/XsdMetadataController.cs:66]: deferred for ODS-4787

File: Application/EdFi.Ods.Features/XsdMetadata/XsdMetadataController.cs [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+
+            var results = schemaInformation.SchemaFiles
+                .Select(
+                    x => new Uri(
+                        $"{_urlHelper.ActionLink("Get", ControllerContext.ActionDescriptor.ControllerName)}/{schemaInformation.SchemaNameMap.UriSegment}/{x}"))
+                .ToList();
+
+            if (!schemaInformation.IsCore())
+            {
+                var coreInformation = await _xsdFileInformationProvider.XsdFileInformationAsync("ed-fi");
```
[the remark this answers is not in the document]
Comment (jamessmckay) [reply Application/EdFi.Ods.Features/XsdMetadata/XsdMetadataController.cs:98]: done

File: Application/EdFi.Ods.Api/Middleware/XsdMetadataFileMiddleware.cs [hunk 13]
```diff
@@ -47,21 +48,23 @@ public async Task InvokeAsync(HttpContext context, RequestDelegate next)
...
+            string assemblyName = xsdFileInformationByUriSegment.AssemblyName;
             string fullQualifiedFileName = $"{assemblyName}.Artifacts.Schemas.{routeValues["file"]}.xsd";
 
             context.Response.ContentType = "application/xml";
             context.Response.StatusCode = StatusCodes.Status200OK;
 
-            await context.Response.WriteAsync(_embeddedFileProvider.File(assemblyName, fullQualifiedFileName));
+            var streamReader = new StreamReader(_embeddedFileProvider.Stream(assemblyName, fullQualifiedFileName));
+
+            await context.Response.WriteAsync( await streamReader.ReadToEndAsync());
```
Comment (blmeyers) [inline Application/EdFi.Ods.Api/Middleware/XsdMetadataFileMiddleware.cs:67]: If you're just going to read to end on the stream, you might as well not have a stream, you're just wrapping more objects around the process of reading the entire file to memory.  If we want to do streams we need to return a stream response.  I think what we're looking for is something like https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.http.sendfileresponseextensions.sendfileasync?view=aspnetcore-3.1?

File: Application/EdFi.Ods.Api/Providers/XsdFileInformationProvider.cs [hunk 14]
```diff
@@ -24,6 +24,7 @@ public class XsdFileInformationProvider : IXsdFileInformationProvider
         private readonly ILog _logger = LogManager.GetLogger(typeof(XsdFileInformationProvider));
         private readonly ISchemaNameMapProvider _schemaNameMapProvider;
         private readonly Lazy<IDictionary<string, XsdFileInformation>> _xsdFileInformationByUriSegment;
+        private const string CoreUriSegment = "ed-fi";
```
Comment (blmeyers) [inline Application/EdFi.Ods.Api/Providers/XsdFileInformationProvider.cs:27]: EdFiConventions.UriSegment is what you want, not a new constant.

**Review by blmeyers — CHANGES_REQUESTED** [review]

File: Application/EdFi.Ods.Api/Middleware/XsdMetadataFileMiddleware.cs (hunk 13 above)
Comment (jamessmckay) [reply Application/EdFi.Ods.Api/Middleware/XsdMetadataFileMiddleware.cs:67]: done

File: Application/EdFi.Ods.Api/Providers/XsdFileInformationProvider.cs (hunk 14 above)
Comment (jamessmckay) [reply Application/EdFi.Ods.Api/Providers/XsdFileInformationProvider.cs:27]: done

**Review by blmeyers — APPROVED** [review]

**Pull request merged**